### PR TITLE
Ian/mysql

### PR DESCRIPTION
**Pull request by edmundhighcock**
This pull request is for a functioning MySQL option. The one TODO is that you have to fill in the connection string on the form, not the individual values like 'host'.

### CODE REVIEW

File: src/server/Generators/classes/mysqlProvider.js [hunk 1]
```diff
@@ -6,29 +8,29 @@ class MySqlProvider {
...
     return conn;
   }
 
   selectWithWhere(table, col, val, returnsMany) {
-    let query = `'SELECT * FROM "${table}" WHERE "${col}" = $1';\n`;
+    let query = `'SELECT * FROM ${table} WHERE ${col} = ?;';\n`;
 
     returnsMany
       ? (query += `${tab.repeat(4)}return connection.query(sql, ${val})\n`)
       : (query += `${tab.repeat(4)}return connection.query(sql, ${val})\n`);
```
Comment (rob2244) [inline src/server/Generators/classes/mysqlProvider.js:25]: Theres no reason to have this conditional check, both routes are the same.
Please change it to:
```javascript 
query += `${tab.repeat(4)}return connection.query(sql, ${val})\n`
```

File: src/server/Generators/classes/mysqlProvider.js [hunk 2]
```diff
@@ -64,20 +66,41 @@ class MySqlProvider {
...
+  }
 }
 
-const addPromiseResolution = () => {
+const addPromiseResolution = (returnsMany=true) => {
   let str = `${tab.repeat(5)}.then(data => {\n`;
-  str += `${tab.repeat(6)}return data;\n`;
+  returnsMany 
+    ?  str += `${tab.repeat(6)}return data[0];\n`
+    :  str += `${tab.repeat(6)}return data[0][0];\n`;
```
Comment (rob2244) [inline src/server/Generators/classes/mysqlProvider.js:100]: Is a multi dimensional array returned when returnsMany is false or is this a typo?

File: src/server/Generators/classes/mysqlProvider.js [hunk 3]
```diff
@@ -6,29 +8,29 @@ class MySqlProvider {
   }
 
   connection() {
-    let conn = `const mysql = require("mysql");\n`;
+    let conn = `const connection = require("mysql-promise");\n`;
```
Comment (rob2244) [inline src/server/Generators/classes/mysqlProvider.js:11]: I would advise against using this library. A quick look at the repo shows that the only function wrapped with promises is the query function, this limits the amount of mileage you can get from the underlying driver if more complex sql is warranted later. I would suggest a library like this instead:
https://www.npmjs.com/package/promise-mysql
it seems more fully featured

File: src/server/Generators/classes/mysqlProvider.js [hunk 4]
```diff
@@ -6,29 +8,29 @@ class MySqlProvider {
...
     conn += `// WARNING - Properly secure the connection string\n`;
-    conn += `const connection = mysql.createConnection(buildMysqlParams(${
+    conn += `connection.configure(${JSON.stringify(mysqlMetadataRetriever.buildMysqlParams(
```
Comment (rob2244) [inline src/server/Generators/classes/mysqlProvider.js:13]: According to the node mysql documentation you can use pass a connection string to connect to the instance as well. Please do it this way to keep the providers as consistent as possible:
[At the bottom of connection options](https://www.npmjs.com/package/mysql#establishing-connections)

File: src/server/Generators/classes/mysqlProvider.js (hunk 3 above)
Comment (rob2244) [reply src/server/Generators/classes/mysqlProvider.js:11]: Additionally, with the library you are currently using the connection object is a function, not an object. The function has to be called to return the object with the query function on it.

E.g.
```javascript
 let conn = `const connection = require("mysql-promise")();\n`;
```

**Review by rob2244 — CHANGES_REQUESTED** [review]
The latest version from dev hasn't been merged into your code, or it hasn't merged correctly. You should be seeing the sql parameter in the pgsql provider methods eg:

```javascript
 select(table) {
    let query = `const sql = 'SELECT * FROM "${table}"';\n`;
    query += `${tab.repeat(4)}return connect.conn.many(sql)\n`;

    query += addPromiseResolution();

    return query;
  }
```

instead of an external reference as in your file: 
```javascript
  select(table) {
    let query = `'SELECT * FROM "${table}"';\n`;
    query += `${tab.repeat(4)}return connect.conn.many(sql)\n`;

    query += addPromiseResolution();

    return query;
  }
```
please merge the latest from dev and then I'll continue my review.
Thank You!

File: src/server/Generators/classes/mysqlProvider.js (hunk 2 above)
Comment (edmundhighcock) [reply src/server/Generators/classes/mysqlProvider.js:100]: Hi! Yes indeed, when you request more than one thing from the underlying mysql library, you will get a multidimensional array, so this is correct.

File: src/server/Generators/classes/mysqlProvider.js (hunk 3 above)
Comment (edmundhighcock) [reply src/server/Generators/classes/mysqlProvider.js:11]: Yes, my bad, I forgot to copy the parentheses from my own local version. Using 'promise-mysql' will be fine, I'm sure... mysql-promise was just the first thing I found that worked!

File: src/server/Generators/classes/mysqlProvider.js (hunk 4 above)
Comment (edmundhighcock) [reply src/server/Generators/classes/mysqlProvider.js:13]: Yes, you can for node mysql, but not for mysql-promise... I'm not sure about promise-mysql.

**Comment by laueian** [comment]
Hey @edmundhighcock, have you made any progress on this PR. Do you still plan to make a PR with the requested changes?

If not we might have someone else make a PR for this feature... just wanted to give you a heads up.

**Comment by laueian** [comment]
Closing out because its been a month and we haven't heard anything from him..

**Comment by edmundhighcock** [comment]
Hi Ian,

I replied directly to the email, but weirdly didn't post here. Thanks for giving me a heads up. To be honest when I submitted the pull request I was expecting you to merge it into ian/mysql. I was happy to work towards merging it with master but I'm short of time at the moment so i can't get round to it quickly. If you want to have someone else work on it or do something different (or scrap it) that's totally cool: I offered the code only in case it was helpful, and I wrote it because I needed a version of your (very nice) tool that worked with mysql at short notice.

Edmund

**Comment by laueian** [comment]
Thanks for the response Edmund! We were really excited to see someone make a PR for the mysql provider. We don't have anyone working on it yet, but we may soon. If you do think you'll have some time to work on it, let us know and we can all work on it together.

I'd be really interested  in learning how you are using SwitchQL and if you have any thoughts about how to improve you. If you have some time this week, I'd love to hop on a call and talk! (We'd really appreciate it! 😄)

**Comment by edmundhighcock** [comment]
I think the tool is really nice. I used it to generate a GraphQL interface to a large database when I was making a sandbox app for a hackathon... it allowed people to browse the database in the sandbox. 

I think what would really improve it would be what we discussed in another thread... it would be great f as well as generating the javascript interface code, you could just click on a 'Deploy' button which would start up a GraphQL server based using the new interface. 

Also, with MySQL, support for more than one (linked) databases on the same server. I was working on this when I ran out of time, and I think it's really just a matter of including the database name as part of the table name.

**Comment by laueian** [comment]
Glad to hear the hackaton went well! Sounds like a cool app.

The two things you mentioned are really great ideas! We actually have some epics (We follow some Agile principles) on our Kanban board in JIRA. I got your email message so we can talk more there.

**Pull request closed without merging**